reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+      // TODO: Detach chunk.
+      req.resolve(CreateIterResultObject(chunk, false));
+
+      if (this._closeRequested === true) {
+        CloseReadableByteStream(stream);
+        return;
+      }
+
+      if (reader._readRequests.length > 0) {
+        CallPullOrPullIntoLaterIfNeeded(this);
+      }
+
+      return;
+    }
+
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    EnqueueInReadableByteStreamController(this, chunk);
+    RespondToReadIntoRequestsFromQueue(this, reader);
+  }
+
+  error(e) {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.error can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._state !== 'readable') {
+      throw new TypeError(`The stream is ${stream._state} and so cannot be errored`);
+    }
+
+    DestroyReadableByteStreamController(this);
+    ErrorReadableByteStream(stream, e);
+  }
+
+  respond(bytesWritten, buffer) {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.respond can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (this._pendingPullIntos.length === 0) {
+      throw new TypeError('No pending BYOB read');
+    }
+
+    assert(IsReadableByteStreamLocked(stream), 'stream must be locked');
+
+    const reader = stream._reader;
+
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    const pullIntoDescriptor = this._pendingPullIntos[0];
+
+    if (pullIntoDescriptor.bytesFilled + bytesWritten > pullIntoDescriptor.byteLength) {
+      throw new RangeError('bytesWritten out of range');
+    }
+
+    if (buffer !== undefined) {
+      pullIntoDescriptor.buffer = buffer;
+    }
+
+    if (stream._state === 'closed') {
+      if (bytesWritten !== 0) {
+        throw new TypeError('bytesWritten must be 0 when calling respond() on a closed stream');
+      }
+
+      assert(pullIntoDescriptor.bytesFilled === 0, 'bytesFilled must be 0');
+
+      while (reader._readIntoRequests.length > 0) {
+        const descriptor = this._pendingPullIntos.shift();
+        RespondToReadIntoRequest(reader, descriptor.buffer);
+      }
+
+      DetachReadableByteStreamReader(reader);
+
+      return;
+    }
+
+    pullIntoDescriptor.bytesFilled += bytesWritten;
+
+    if (pullIntoDescriptor.bytesFilled >= pullIntoDescriptor.elementSize) {
+      this._pendingPullIntos.shift();
+
+      let remainder;
+      if (pullIntoDescriptor.bytesFilled % pullIntoDescriptor.elementSize > 0) {
+        remainder = buffer.slice(result.bytesUsed, pullIntoDescriptor.bytesFilled);
+      }
+
+      RespondToReadIntoRequest(reader, pullIntoDescriptor.buffer, pullIntoDescriptor.bytesFilled);
```

Comment:
This code is not tested and actually doesn't work. I've fixed it and made it tested by 28cbdc8bd875ce5a1800fa9a38e0e2178fef887c.